### PR TITLE
Add routes support for V3

### PR DESCRIPTION
**Pull request by enchobelezirev**
Implementation of the Routes API(https://v3-apidocs.cloudfoundry.org/version/3.77.0/index.html#routes) in the client.

### CODE REVIEW

File: cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/applications/ReactorApplicationsV3.java [hunk 1]
```diff
@@ -16,49 +16,8 @@
...
-import org.cloudfoundry.client.v3.applications.StartApplicationResponse;
-import org.cloudfoundry.client.v3.applications.StopApplicationRequest;
-import org.cloudfoundry.client.v3.applications.StopApplicationResponse;
-import org.cloudfoundry.client.v3.applications.TerminateApplicationInstanceRequest;
-import org.cloudfoundry.client.v3.applications.UpdateApplicationEnvironmentVariablesRequest;
-import org.cloudfoundry.client.v3.applications.UpdateApplicationEnvironmentVariablesResponse;
-import org.cloudfoundry.client.v3.applications.UpdateApplicationRequest;
-import org.cloudfoundry.client.v3.applications.UpdateApplicationResponse;
+import org.cloudfoundry.client.v3.applications.ListApplicationRoutesRequest;
+import org.cloudfoundry.client.v3.applications.*;
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/applications/ReactorApplicationsV3.java]: House style is to specify all imports, no wildcards.

File: cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/spaces/ReactorSpacesV3.java [hunk 2]
```diff
@@ -16,17 +16,7 @@
...
-import org.cloudfoundry.client.v3.spaces.CreateSpaceRequest;
-import org.cloudfoundry.client.v3.spaces.CreateSpaceResponse;
-import org.cloudfoundry.client.v3.spaces.GetSpaceIsolationSegmentRequest;
-import org.cloudfoundry.client.v3.spaces.GetSpaceIsolationSegmentResponse;
-import org.cloudfoundry.client.v3.spaces.GetSpaceRequest;
-import org.cloudfoundry.client.v3.spaces.GetSpaceResponse;
-import org.cloudfoundry.client.v3.spaces.ListSpacesRequest;
-import org.cloudfoundry.client.v3.spaces.ListSpacesResponse;
-import org.cloudfoundry.client.v3.spaces.SpacesV3;
+import org.cloudfoundry.client.v3.spaces.*;
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/spaces/ReactorSpacesV3.java]: House style is to specify all imports, no wildcards.

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/applications/_ListApplicationRoutesRequest.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+     * The organization guids
+     */
+    @FilterParameter("organization_guids")
+    abstract List<String> getOrganizationGuids();
+
+    /**
+     * The metadata query
+     */
+    @FilterParameter("label_selector")
+    @Nullable
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/applications/_ListApplicationRoutesRequest.java]: FilterParameter has an implicit Nullable so this shouldn't be necessary. Having said that I've noticed I've done it in one or two places, so either we've both made the same mistake or there's a reason for this that I've forgotten. Can you clarify?

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/applications/_ListApplicationRoutesRequest.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    abstract List<String> getPaths();
+
+
+    /**
+     * The domain guids
+     */
+    @FilterParameter("domain_guids")
+    abstract List<String> getDomainGuids();
+
+
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/applications/_ListApplicationRoutesRequest.java]: Remove extra line

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/routes/Route.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+     * The relationships
+     */
+    @JsonProperty("relationships")
+    public abstract RouteRelationships getRelationships();
+
+    /**
+     * The metadata
+     */
+    @JsonProperty("metadata")
+    @Nullable
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/routes/Route.java]: See earlier Nullable comment

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/routes/RoutesV3.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+     */
+    Mono<UpdateRouteResponse> update(UpdateRouteRequest request);
+
+    /**
+     * Makes the <a href="https://v3-apidocs.cloudfoundry.org/version/3.77.0/index.html#delete-route">Delete Route</a> request
+     *
+     * @param request the Delete Route request
+     * @return the response from Delete Route request
+     */
+    Mono<String> delete(DeleteRouteRequest request);
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/routes/RoutesV3.java]: Methods should be in alphabetical order

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/routes/_ListRoutesRequest.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+     * The organization guids
+     */
+    @FilterParameter("organization_guids")
+    abstract List<String> getOrganizationGuids();
+
+    /**
+     * The metadata query
+     */
+    @FilterParameter("label_selector")
+    @Nullable
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/routes/_ListRoutesRequest.java]: see earlier Nullable comment

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/routes/_CreateRouteRequest.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+@JsonSerialize
+@Value.Immutable
+public abstract class _CreateRouteRequest {
+
+    /**
+     * The host
+     */
+    @Nullable
+    @JsonProperty("host")
+    public abstract String getHost();
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/routes/_CreateRouteRequest.java]: Public shouldn't be needed here

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/routes/_UpdateRouteRequest.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+@JsonSerialize
+@Value.Immutable
+public abstract class _UpdateRouteRequest {
+
+    /**
+     * The route id
+     */
+    @JsonIgnore
+    public abstract String getRouteId();
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/routes/_UpdateRouteRequest.java]: Public shouldn't be needed here

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/spaces/_DeleteUnmappedRoutesRequest.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.util.List;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import org.cloudfoundry.client.v3.FilterParameter;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public abstract  class _DeleteUnmappedRoutesRequest {
+    /**
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/spaces/_DeleteUnmappedRoutesRequest.java]: Add blank line here

File: cloudfoundry-client/src/test/java/org/cloudfoundry/client/v3/routes/CreateRouteRequestTest.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+import org.cloudfoundry.client.v3.Relationship;
+import org.cloudfoundry.client.v3.ToOneRelationship;
+import org.junit.Rule;
+import org.junit.Test;
+
+public final class CreateRouteRequestTest {
+
+    @Test(expected = IllegalStateException.class)
+    public void invalid() {
```
Comment (twoseat) [inline cloudfoundry-client/src/test/java/org/cloudfoundry/client/v3/routes/CreateRouteRequestTest.java]: Should test for each missing field (seems overly detailed, but I've found errors this way in the past)

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/applications/_ListApplicationRoutesRequest.java (hunk 3 above)
Comment (enchobelezirev) [reply cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/applications/_ListApplicationRoutesRequest.java]: I have looked how to make the POJOs from the already existing places :) So, if it is mistake, I could remove the @Nullable from everywhere

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/routes/Route.java (hunk 5 above)
Comment (enchobelezirev) [reply cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/routes/Route.java]: Is this applicable for the @JsonProperty annotation?

Comment (twoseat) [reply cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/routes/Route.java]: You're right, I'm wrong, ignore this one!

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/applications/_ListApplicationRoutesRequest.java (hunk 3 above)
Comment (twoseat) [reply cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/applications/_ListApplicationRoutesRequest.java]: I think remove it for now, if it causes problems it should turn up in integration testing and I'll fix it then. I did just notice that the javadoc you have is wrong for this though.

**Comment by enchobelezirev** [comment]
Hello @twoseat,
I have created a commit with the Organised imports and the fixed comments.
The thing about @Nullable on non-collection variable does not work - I have tested it.
Could you review the PR once again and tell me whether it is fine? Thanks!

**Comment by aegershman** [comment]
any thoughts on if/when v3 routes support could be incorporated? is it in a holding pattern at the moment?

**Comment by twoseat** [comment]
@aegershman We're currently having a push on v3 implementation, so while I don't have an exact date I'd expect it to be in the next few weeks as I'm focussing first on areas where we have PRs such as this.

**Pull request closed without merging**